scipy/integrate/_cubature.py

```diff
@@ -328,22 +733,465 @@ def cubature(f, a, b, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
 
         status = "converged" if success else "not_converged"
 
+        # Apply sign change to handle any limits which were initially flipped.
+        est = sign * est
+
         return CubatureResult(
             estimate=est,
             error=err,
             status=status,
             subdivisions=subdivisions,
-            regions=regions,
+            patches=patches,
             atol=atol,
             rtol=rtol,
         )
 
 
-def _process_subregion(data):
+def _process_subpatch(data):
     f, rule, args, coord = data
     a_k_sub, b_k_sub = coord
 
     est_sub = rule.estimate(f, a_k_sub, b_k_sub, args)
     err_sub = rule.estimate_error(f, a_k_sub, b_k_sub, args)
 
     return a_k_sub, b_k_sub, est_sub, err_sub
+
+
+def _is_strictly_in_patch(point, a, b):
+    xp = array_namespace(point, a, b)
+
+    if xp.all(point == a) or xp.all(point == b):
+        return False
+
+    return xp.all(a <= point) and xp.all(point <= b)
+
+
+def _split_patch_at_points(a, b, points):
+    """
+    Given the integration limits `a` and `b` describing a rectangular patch and a list
+    of `points`, find the list of ``[(a_1, b_1), ..., (a_l, b_l)]`` which breaks up the
+    initial patch into smaller subpatches such that no `points` lie strictly inside
+    any of the subpatches.
+    """
+
+    xp = array_namespace(a, b)
+    patches = [(a, b)]
+
+    for point in points:
+        if xp.any(xp.isinf(point)):
+            # If a point is specified at infinity, ignore.
+            #
+            # This case occurs when points are given by the user to avoid, but after
+            # applying a transformation, they are removed.
+            continue
+
+        new_subpatches = []
+
+        for a_k, b_k in patches:
+            if _is_strictly_in_patch(point, a_k, b_k):
+                subpatches = _split_patch(a_k, b_k, point)
+
+                for left, right in subpatches:
+                    # Skip any zero-width patches.
+                    if xp.any(left == right):
+                        continue
+                    else:
+                        new_subpatches.append((left, right))
+
+                new_subpatches.extend(subpatches)
+
+            else:
+                new_subpatches.append((a_k, b_k))
+
+        patches = new_subpatches
+
+    return patches
+
+
+class _VariableTransform:
+    """
+    A transformation that can be applied to an integral.
+    """
+
+    @property
+    def transformed_limits(self):
+        """
+        New limits of integration after applying the transformation.
+        """
+
+        raise NotImplementedError
```

Comment:
I don't think it matters a lot here, but I think [there may be some benefits to making this an abstract class](https://dbader.org/blog/abstract-base-classes-in-python) like:

```python
from abs import ABC, abstractmethod
...
class _VariableTransform(ABC):
...
    @property
    @abstractmethod
    def transformed_limits(self):
        """
        New limits of integration after applying the transformation.
        """
```

When a docstring is included in an `abstractmethod`, there is no need to `raise NotImplementedError` or add anything else like `pass` or `...`.

(https://peps.python.org/pep-3119/#the-abc-module-an-abc-support-framework)

Reply:
IIRC the suggestion to not use the ABC module was mine. Again IIRC, the mro gets a bit weird when inheriting from abc etc.